File: sdk/telemetry/go/pda_test.go
package telemetry

import (
	"testing"

	"github.com/gagliardetto/solana-go"
)

var testProgramID = solana.MustPublicKeyFromBase58("tE1exJ5VMyoC9ByZeSmgtNzJCFF74G9JAv338sJiqkC")

func TestDeriveDeviceLatencySamplesPDA(t *testing.T) {
	origin := solana.MustPublicKeyFromBase58("11111111111111111111111111111112")
	target := solana.MustPublicKeyFromBase58("11111111111111111111111111111113")
	link := solana.MustPublicKeyFromBase58("11111111111111111111111111111114")

	addr, bump, err := DeriveDeviceLatencySamplesPDA(testProgramID, origin, target, link, 42)
	if err != nil {
		t.Fatalf("DeriveDeviceLatencySamplesPDA: %v", err)
	}
	if addr.IsZero() {
		t.Error("derived zero address")
	}

	addr2, bump2, err := DeriveDeviceLatencySamplesPDA(testProgramID, origin, target, link, 42)
	if err != nil {
		t.Fatalf("DeriveDeviceLatencySamplesPDA (2nd): %v", err)
	}
	if addr != addr2 || bump != bump2 {
		t.Error("PDA derivation not deterministic")
	}
}

func TestDeriveInternetLatencySamplesPDA(t *testing.T) {
	oracle := solana.MustPublicKeyFromBase58("11111111111111111111111111111112")
	origin := solana.MustPublicKeyFromBase58("11111111111111111111111111111113")
	target := solana.MustPublicKeyFromBase58("11111111111111111111111111111114")

	addr, _, err := DeriveInternetLatencySamplesPDA(testProgramID, oracle, "RIPE Atlas", origin, target, 42)
	if err != nil {
		t.Fatalf("DeriveInternetLatencySamplesPDA: %v", err)
	}
	if addr.IsZero() {
		t.Error("derived zero address")
	}
}

File: sdk/telemetry/typescript/telemetry/tests/pda.test.ts
/**
 * PDA derivation tests.
 */

import { describe, expect, test } from "bun:test";
import { PublicKey } from "@solana/web3.js";
import {
  deriveDeviceLatencySamplesPda,
  deriveInternetLatencySamplesPda,
} from "../pda.js";

const PROGRAM_ID = new PublicKey(
  "tE1exJ5VMyoC9ByZeSmgtNzJCFF74G9JAv338sJiqkC",
);

describe("deriveDeviceLatencySamplesPda", () => {
  test("deterministic", () => {
    const origin = new PublicKey("11111111111111111111111111111112");
    const target = new PublicKey("11111111111111111111111111111113");
    const link = new PublicKey("11111111111111111111111111111114");

    const [addr1, bump1] = deriveDeviceLatencySamplesPda(
      PROGRAM_ID,
      origin,
      target,
      link,
      42,
    );
    const [addr2, bump2] = deriveDeviceLatencySamplesPda(
      PROGRAM_ID,
      origin,
      target,
      link,
      42,
    );

    expect(addr1.equals(addr2)).toBe(true);
    expect(bump1).toBe(bump2);
  });
});

describe("deriveInternetLatencySamplesPda", () => {
  test("non-zero", () => {
    const oracle = new PublicKey("11111111111111111111111111111112");
    const origin = new PublicKey("11111111111111111111111111111113");
    const target = new PublicKey("11111111111111111111111111111114");

    const [addr] = deriveInternetLatencySamplesPda(
      PROGRAM_ID,
      oracle,
      "RIPE Atlas",
      origin,
      target,
      42,
    );
    expect(addr.equals(PublicKey.default)).toBe(false);
  });
});
